machine/v1beta1/types_alibabaprovider.go

```diff
@@ -0,0 +1,398 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta1
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// AlibabaCloudMachineProviderConfig is the Schema for the alibabacloudmachineproviderconfig API
+// Compatibility level 2: Stable within a major release for a minimum of 9 months or 3 minor releases (whichever is longer).
+// +openshift:compatibility-gen:level=2
+// +k8s:openapi-gen=true
+type AlibabaCloudMachineProviderConfig struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// More detail about alibabacloud ECS
+	// https://www.alibabacloud.com/help/doc-detail/25499.htm?spm=a2c63.l28256.b99.727.496d7453jF7Moz
+
+	//The instance type of the instance.
+	InstanceType string `json:"instanceType"`
+
+	// The ID of the vpc
+	// +optional
+	VpcID string `json:"vpcId,omitempty"`
+
+	// The ID of the region in which to create the instance. You can call the DescribeRegions operation to query the most recent region list.
+	RegionID string `json:"regionId"`
+
+	// The ID of the zone in which to create the instance. You can call the DescribeZones operation to query the most recent region list.
+	ZoneID string `json:"zoneId"`
+
+	// The ID of the image used to create the instance.
+	ImageID string `json:"imageId"`
+
+	// DataDisk hold information regarding the extra data disks
+	// +optional
+	DataDisks []DataDiskProperties `json:"dataDisk,omitempty"`
+
+	// SecurityGroups is an array of references to security groups which to assign the instance. The valid values of N vary based on the
```

Comment:
Just trying to improve the readability of this one. What's the N in this, I don't think this is that readable, can we reword this at all?
```suggestion
	// SecurityGroups is list of references to security groups to assign to the instance. The valid values of N vary based on the
```